src/NHSD.BuyingCatalogue.Identity.Api/Controllers/AccountController.cs

```diff
@@ -35,33 +53,61 @@ public IActionResult Login(string returnUrl = "/")
         [ValidateAntiForgeryToken]
         public async Task<IActionResult> Login(LoginViewModel viewModel)
         {
-            viewModel.ThrowIfNull();
+            viewModel.ThrowIfNull(nameof(viewModel));
+
+            var returnUrl = viewModel.ReturnUrl.ToString();
 
-            var user = TestUsers.Users
-                .FirstOrDefault(usr => usr.Password == viewModel.Password && usr.Username == viewModel.Username);
+            // Check if we are in the context of an authorization request
+            var context = await _interaction.GetAuthorizationContextAsync(returnUrl);
 
-            if (user is null)
+            LoginViewModel NewLoginViewModel() =>
+                new LoginViewModel { ReturnUrl = viewModel.ReturnUrl, Username = context?.LoginHint };
+
+            if (!ModelState.IsValid)
+                return View(NewLoginViewModel());
+
+            var result = await _signInManager.PasswordSignInAsync(viewModel.Username, viewModel.Password, false, true);
+            if (!result.Succeeded)
             {
-                return Unauthorized();
+                await _events.RaiseAsync(new UserLoginFailureEvent(viewModel.Username, "invalid credentials", clientId: context?.ClientId));
+                ModelState.AddModelError(string.Empty, "Invalid username or password");
+
+                return View(NewLoginViewModel());
             }
 
-            await HttpContext.SignInAsync(user.SubjectId, user.Username);
+            var user = await _userManager.FindByNameAsync(viewModel.Username);
+            await _events.RaiseAsync(new UserLoginSuccessEvent(user.UserName, user.Id, user.UserName, clientId: context?.ClientId));
 
-            string returnUrl = viewModel.ReturnUrl;
-            if (_interaction.IsValidReturnUrl(returnUrl))
+            if (context != null)
             {
+                if (await _clientStore.IsPkceClientAsync(context.ClientId))
+                {
+                    // If the client is PKCE then we assume it's native, so this change in how to
+                    // return the response is for better UX for the end user.
+                    return View("Redirect", new RedirectViewModel(viewModel.ReturnUrl));
+                }
+
+                // We can trust viewModel.ReturnUrl since GetAuthorizationContextAsync returned non-null
                 return Redirect(returnUrl);
             }
 
-            return Redirect("~/");
+            // Request for a local page
+            if (Url.IsLocalUrl(returnUrl))
+                return Redirect(returnUrl);
```

Comment:
In some Identity examples I have seen the use of a LocalRedirect method. This should ensure that the return URL is a route actually within the application. I think this would make the conditional check of IsLocalUrl redundant.